### ISSUE TITLE
Migrate the parser to the language definition v2

### ISSUE THREAD

**Issue by Xanewok** [NomicFoundation/slang#638]
At the moment we use our Rust macros 1.0 (aka DSL v1) to define the language for our parser. We already have:
1. YAML spec
2. DSL v1
3. DSL v2

So to cut down the different definitions, we should migrate to the newest definition (see https://github.com/NomicFoundation/slang/pull/617).

Migration PR:

- [X] #650

Impacts the AST:

- [X] https://github.com/NomicFoundation/slang/pull/693
- [X] https://github.com/NomicFoundation/slang/pull/698
- [X] https://github.com/NomicFoundation/slang/pull/699
  - Clean up `PrecedenceExpression#rule_name` and adapt the codegen model to fit the v2 definition better
  - `Precedence` items should also produce their own `NonTerminalKind` kind, wrapping the single child (`PrecedenceExpression` or `PrimaryExpression`).
  - `PrecedenceExpression` should get their own `parse_XXX()` function.

Does not impact the AST:

- ~[ ] Re-introduce EOF/EOI (End of Input) virtual token/trivia parser~
- ~[ ] Change LeadingTrivia definition in v2 to allow being terminated by EOI as well (changed in #650)~ (tracked in #1020)
- [x] Don't monkey-patch SourceUnit rule to always be followed by EndOfFileTrivia (copies v1; trivia model in v2 is a bit different)
- [x] Always try to parse `ZeroOrMore(AnyTriviaItem)` at the end when parsing any production with `pub fn parse`
- [x] Stop leaking identifiers by adapting either v1 or v2 models (#977)
- [x] Keyword trie inclusion should be reworked to not require synthetic rules over all keywords (v1 model) (#1018)
- [x] Revert https://github.com/NomicFoundation/slang/pull/646 (done in #800)
- [x] Ordering of parsing statements variants ([context](https://github.com/NomicFoundation/slang/pull/653#discussion_r1393651832)).~ (separated into #1019)
- [x] See if we can remove `Scanner::not_followed_by` completely and automatically calculate it instead ([context](https://github.com/NomicFoundation/slang/pull/650#discussion_r1408406775)).
- [x] Remove redundant versioning checks in precedence expressions ([context](https://github.com/NomicFoundation/slang/pull/650#discussion_r1406913328)).
- [x] Precedence individual operators need to be deduped across versions ([context](https://github.com/NomicFoundation/slang/pull/687#discussion_r1413111706)).
- [X] Delete `codegen_grammar` crate. (https://github.com/NomicFoundation/slang/pull/972)

**Comment by Xanewok** [comment]
While implementing #1002 I started thinking whether using v2 types *only* makes sense for our existing parser generator.

The goal of #1002 and this issue in general was to reduce duplication between the very similar types and get rid of the bridging types whenever possible. 

While it was enabled by the mechanical and lexer-oriented design of `model::Scanner`, I'm worried that it might not be possible to do the same with `ParserDefinitionNode` because of the higher level of abstraction of the v2 types for structs, choices and precedence operators.

The are a couple of points here:
1. The v1 model for the parser definition was very generic and dynamic in a good way - I feel that trying to shoehorn the v2 to fit into it will make it harder to reason about and make it less composable
2. Moreover, because of the mutually recursive nature of the parser (and how v2 inherently differs from v1: flat list of definitions vs recursively defined parsers via DSL macro) we need to resolve/lower the v2 definition anyway into a structured representation
3. The structured representation is required to perform some analysis for the purposes of codegen, like reachable terminals in a given lexical context etc.

Having the above in mind, while there are some unnecessary translation hacks still that I'd like to smooth out, I would probably reconsider dropping the v1 interface entirely.

cc @AntonyBlakey @OmarTawfik for visibility

**Comment by Xanewok** [comment]
@OmarTawfik from what I understand:
1. we don't need the end of input token (first item) and
2. the "Change LeadingTrivia definition in v2 to allow being terminated by EOI as well" (second item) is somewhat subsumed by us trying to always parse the trivia after the end of the production.

If not, then, I'd propose that we treat it as general improvement to the trivia handling rather than being a core component of the DSL v2 (we did already migrate the parser to the DSL v2 and we're happy with the current behavior) and we should probably track it separately if we're not already happy with what we have, as per above.

The listed scope for this issue almost feels like a quarterly KR and I'd love if we could finally close it, once #1018 is merged (and we revert the string rule) :sweat_smile:

**Comment by OmarTawfik** [comment]
> we don't need the end of input token (first item)

sounds good.

> is somewhat subsumed by us trying to always parse the trivia after the end of the production.

IIRC, the hack we added is to "reuse" leading trivia as EOF trivia, and the proper fix was to construct a specific EOF parser in codegen, with its definition being `Choice([all trivia items in the grammar])` and run it after each successful parse. This allows `LeadingTrivia` definition to remain true/accurate to DSL v2, and remove the `TODO(#638)` comments in the definition.

Is that still the case? If so, it is definitely not blocking, and I would not prioritize it over other important tasks we have, but I think we should still do it. If you have a full plate, please feel free to put it in a separate issue and assign it to me, and we can close this one.

**Comment by Xanewok** [comment]
I've opened #1020 to track this.

**Comment by OmarTawfik** [comment]
>  If you're fine with it, I'd propose that we separate an issue/thread dedicated to it as I feel we had already too many items bundled in this issue and the decision is not clear-cut at this moment.

Sounds good. Let's follow up on #1020. Thanks!
